### PR TITLE
fix(agent): require bounded curator peer discovery pages

### LINKED ISSUE #2062: perf(discovery): make curator peer pagination a required bounded capability
## Summary

#2053 adds a production `DiscoveryClient.findAgentPeerIdsByAddress()` keyset query so VM exact recovery can walk an oversized curator wallet's peer registry without retaining a full rich-profile result. The resolver currently detects that method dynamically and falls back to unbounded `findAgents({ agentAddress })` when a custom/older discovery implementation does not expose it.

That compatibility fallback is correct for existing integrations, but it leaves an avoidable scalability cliff: `findAgents` selects OPTIONAL profile fields that can multiply rows before deduplication and materializes the complete result before the caller applies its bounded roster cap. An alternative discovery adapter can therefore turn a bounded recovery pass back into an unbounded registry query and heap allocation.

## Current behavior

- Production `DiscoveryClient` supports ordered keyset paging by `peerId`.
- `resolveCuratorPeerIdsForCg()` uses it only when `typeof findAgentPeerIdsByAddress === 'function'` and `maxPeerIds` is supplied.
- Otherwise the resolver calls rich `findAgents`, collects every row, then deduplicates/sorts peer IDs in memory.
- Recovery intentionally fails open on rosters larger than the absence-proof cap, so bounded page transport is sufficient; it never needs the entire oversized registry in one allocation.

## Recommended fix direction

1. Promote wallet-to-peer keyset pagination into the typed discovery capability used by the agent instead of feature-detecting an optional method.
2. Define explicit page semantics: stable `ORDER BY peerId`, exclusive `afterPeerId`, bounded positive `limit`, duplicate-free peer IDs, abort propagation, and a clear end/continuation signal.
3. Adapt all production discovery providers to that contract.
4. For a provider that cannot paginate, return an explicit unsupported/lookup-failed result; do not silently run an unbounded rich-profile query from a bounded recovery call.
5. Keep legacy single-curator triple resolution unchanged for non-wallet-scoped graph IDs.
6. Do not add a registry cache, background crawler, timer, or durable cursor solely for this path.

## Acceptance criteria

- [ ] Every bounded curator resolution performs at most `limit + 1` result-row retention per page.
- [ ] OPTIONAL profile multiplicity cannot hide a peer or inflate recovery memory.
- [ ] An adapter without pagination fails closed/fail-open explicitly and never falls back to an unbounded full registry materialization.
- [ ] Keyset pages are duplicate-free, monotonic, abortable, and eventually cover a stable registry.
- [ ] Insert/delete churn between pages cannot authorize a false complete-absence proof; overflow remains unconfirmed.
- [ ] Production, injected, and test discovery implementations satisfy the same typed contract.
- [ ] Resource tests cover a wallet with more peers than the proof cap and profiles with many OPTIONAL values.

## Related

- Parent pressure issue: #2052
- Partial implementation: #2053
- General agent listing pagination: #310

### PR DESCRIPTION
**Pull request by branarakic**
Bounded curator recovery could fall back to `findAgents()` when a provider lacked peer pagination, loading the entire rich-profile registry before applying its roster cap. Its metadata-refresh fallback could repeat the same unbounded query.

Both paths now require the typed `AgentPeerDiscovery` page contract. Missing, unsupported, or malformed providers produce a lookup failure; they cannot trigger rich-profile fallback. `DiscoveryClient.findAgentPeerPageByAddress()` requires a positive page size capped at 1,024, retains at most one lookahead row, and returns an explicit continuation. Peer IDs are normalized in the SPARQL DISTINCT projection before LIMIT, so optional-profile multiplicity and alternate RDF encodings cannot consume page slots or hide later peers. Cancellation is forwarded and rechecked after provider/query completion.

The recovery owner consumes explicit complete, continue, and cycle traversal states. Tail pages remain incomplete roster evidence even when registry churn makes them short; cycle clears the saved cursor before the next recovery pass. The existing SDK overflowed and nextPageAfterPeerId fields remain available as compatibility projections, including support for custom providers returning the prior shape. Exhausted cursors wrap through a fresh bounded first-page query. Legacy single-curator metadata resolution and explicit rich-profile calls keep their existing behavior. The published `findAgentPeerIdsByAddress(wallet, options?)` method remains as a deprecated, array-returning compatibility facade that walks bounded pages internally. Bounded recovery requires the new page method and never invokes this facade. The README documents both contracts.

Closes #2062.

Validation:
- 345 tests pass across 11 discovery, contract, planning, recovery-consumer, metadata-refresh, and telemetry suites.
- The real embedded-Oxigraph resource fixture walks 273 peers (above the 256-peer proof cap), 40 optional framework values per profile, and duplicate peer-ID RDF encodings. Seven-peer pages return at most eight query rows, with no duplicated or omitted peers.
- Restoring the prior canary lifecycle implementation makes the missing-pagination regression fail. The working file was restored byte-for-byte before final validation.
- Three regressions fail against the preceding implementation and pass after the fix: a provider-supplied array iterator, exhausted-cursor restart, and legacy-shaped partial rosters being incorrectly confirmed and put into recovery backoff. Both legacy and explicit traversal owners visit all five peers and return to a newly populated first peer without suppressing recovery.
- A successful bounded wallet-curator fallback returns the registry peer with registry provenance, requests limit 1, avoids rich discovery, and overrides a bootstrap hint without earning authority.
- Public runtime/type compatibility checks preserve omitted-limit, limited, and cursor-based array calls, including limits larger than one page.
- Agent build, public type fixtures, strict changed runtime-fixture types, lint, the 1,801-file test inventory, SPARQL checks, and merge compatibility pass. The two discovery suites are explicitly included in the agent unit lane.
- The page-contract and bounded roster traversal modules each have 100% scoped statement/branch/function/line coverage. This is not a whole-package coverage claim; no live chain or remote production deployment was exercised.

### CODE REVIEW

File: packages/agent/src/agent-peer-discovery.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    throw new RangeError(`Peer page limit must be an integer from 1 to ${MAX_AGENT_PEER_PAGE_SIZE}`);
+  }
+  if (request.afterPeerId !== undefined && (typeof request.afterPeerId !== 'string' || request.afterPeerId.length === 0)) {
+    throw new TypeError('Peer page cursor must be a nonempty string');
+  }
+  request.signal?.throwIfAborted();
+}
+
+/** Reject a broken provider contract before treating a page as registry evidence. */
+export function validateAgentPeerPage(page: AgentPeerPage, request: AgentPeerPageRequest): AgentPeerPage {
```
Comment (otReviewAgent) [inline packages/agent/src/agent-peer-discovery.ts]: 🟡 Issue: Model provider output as untrusted until validation completes

**What's wrong**
The validator's type signature claims malformed external output is already valid. This obscures the actual trust boundary, forces casts in tests, and makes future additions to the page contract easy to validate incompletely.

**Example**
A provider returning `{ peerIds: ['peer-001'] }` must be cast to `AgentPeerPage` before this validator can be tested, despite lacking the required continuation field.

**Suggested direction**
Make `validateAgentPeerPage` a genuine decoder from `unknown` to `AgentPeerPage`, rather than asserting a contract in the parameter type and checking it again at runtime.

**For Agents**
In `agent-peer-discovery.ts`, accept `unknown` at the provider-response validation boundary, narrow the object and both fields explicitly, then return `AgentPeerPage`. Preserve the monotonicity, bound, cancellation, and defensive-copy behavior; update malformed-provider tests so they need no cast.

File: packages/agent/src/dkg-agent-cg-resolve.ts [hunk 2]
```diff
@@ -657,14 +658,17 @@ export async function resolveCuratorSyncPeer(
     if (!resolved) {
       try {
         throwIfSyncAuthAborted(options.signal);
-        const agents = await agent.discovery.findAgents({ signal: options.signal });
+        const peerId = options.registryPageLimit === undefined
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-cg-resolve.ts]: 🟡 Issue: Keep registry lookup policy out of metadata-refresh plumbing

**What's wrong**
A low-level pagination parameter now leaks through the generic refresh API solely to select bounded versus legacy lookup behavior. This couples metadata refresh to registry implementation details and creates a fragile optional mode that every intermediate layer must preserve correctly.

**Example**
A bounded roster lookup calls `refreshMetaFromCurator(..., { registryPageLimit: 1 })`; refresh forwards that option into curator resolution, which then branches between `findAgents()` and `readAgentPeerPage()`. Understanding whether refresh is bounded now requires tracing three modules.

**Suggested direction**
Give bounded and legacy curator resolution explicit entry points, or inject the peer-resolution operation into refresh. Keep page sizing owned by the discovery/recovery layer.

**For Agents**
Refactor `curator-meta-refresh.ts`, `dkg-agent-cg-resolve.ts`, and the lifecycle caller so metadata refresh receives a resolved peer or a named resolver/policy rather than a registry page-size sentinel. Preserve the legacy rich-profile route for legacy callers and prove that bounded recovery never invokes `findAgents()`.

File: packages/agent/src/dkg-agent-lifecycle.ts [hunk 3]
```diff
@@ -7311,29 +7312,30 @@ export class LifecycleSyncMethods extends DKGAgentBase {
...
+              page = await queryPage();
             }
             assertCurrent();
-            const overflowed = !pageStartedAtBeginning
-              || peerIds.length > options.maxPeerIds;
-            const bounded = peerIds.slice(0, overflowed ? pagePeerIds : options.maxPeerIds);
+            // Only a complete first-page query can describe a whole roster.
+            // A tail page remains overflowed even if concurrent deletions made
+            // it short; it cannot authorize an absence proof for earlier peers.
+            const overflowed = !pageStartedAtBeginning || page.nextAfterPeerId !== null;
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-lifecycle.ts]: 🟡 Issue: Represent roster traversal states explicitly instead of overloading the cursor

**What's wrong**
The new flow uses the same `overflowed`/cursor pair for both “the registry has another page” and “this tail page cannot prove completeness, so cycle later.” Those states require different orchestration, producing a hidden empty-page restart protocol that is difficult for callers and maintainers to infer.

**Example**
For `{ peerIds: ['peer-004'], nextAfterPeerId: null }` after a cursor, the result becomes `{ overflowed: true, nextPageAfterPeerId: 'peer-004' }`. The next call queries after `peer-004`, receives an empty page, and silently restarts at page one.

**Suggested direction**
Return a discriminated traversal result such as `complete`, `continue`, or `cycle`, and let the recovery owner decide how to persist/advance it. This removes the mutable `pageStartedAtBeginning` protocol and the synthesized continuation semantics from the lifecycle god-file.

**For Agents**
Extract the bounded curator-roster traversal from `resolveCuratorPeerIdsForCg` into a focused module. Represent complete, continue, and cycle/restart outcomes explicitly, while preserving the current churn safety, one-peer rotation, cancellation, and exhausted-cursor behavior. Keep the existing tail-page and wrap tests against that abstraction.

File: packages/agent/src/dkg-agent-cg-resolve.ts [hunk 4]
```diff
@@ -657,14 +658,17 @@ export async function resolveCuratorSyncPeer(
...
+          ? (await agent.discovery.findAgents({ signal: options.signal })).find(
+              (candidate) => candidate.agentAddress?.toLowerCase() === curatorIdentifier.toLowerCase(),
+            )?.peerId
+          : (await readAgentPeerPage(agent.discovery, curatorIdentifier, {
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-cg-resolve.ts]: 🟡 Issue: The successful bounded registry fallback is not tested

**What's wrong**
The tests prove that an empty bounded lookup fails closed and avoids the rich-profile fallback, but they never prove that a valid bounded result is actually used to resolve the curator. A regression that discarded `peerIds[0]`, failed to propagate it, or always returned the bootstrap hint would leave bounded metadata recovery unable to contact a discoverable curator while the suite remained green.

**Example**
Add a resolver test where metadata names a wallet curator, `findAgentPeerPageByAddress` returns `{ peerIds: ['peer-curator'], nextAfterPeerId: null }`, and `findAgents` throws. Calling `resolveCuratorSyncPeer(..., { registryPageLimit: 1 })` should return `peer-curator` with `registry` provenance and issue a page request with limit 1.

**Suggested direction**
Add a positive integration-level resolver test for the new bounded branch, complementing the existing empty-page/failure-path test.

**For Agents**
In `sync-policy.test.ts` or `bounded-curator-discovery.test.ts`, exercise `resolveCuratorSyncPeer`/`resolveCuratorPeerId` with a wallet curator, `registryPageLimit`, and a valid non-empty page. Assert the peer and provenance, the requested bound, and that `findAgents()` is not called.

File: packages/agent/src/agent-peer-discovery.ts (hunk 1 above)
Comment (branarakic-agent) [reply packages/agent/src/agent-peer-discovery.ts]: Addressed in 04dd6272c. validateAgentPeerPage now accepts unknown, rejects non-object/array/null inputs before reading fields, narrows peerIds and nextAfterPeerId explicitly, and returns a fresh typed AgentPeerPage. The invalid-provider tests now pass raw malformed values without AgentPeerPage casts.

File: packages/agent/src/dkg-agent-cg-resolve.ts (hunk 2 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-cg-resolve.ts]: Addressed in 04dd6272c. The numeric registryPageLimit sentinel is removed from generic metadata refresh. Refresh now accepts an explicit caller-owned curatorPeerResolver; bounded recovery supplies the named bounded-first-page registry policy, while ordinary/legacy callers retain findAgents. The bounded path is covered to ensure findAgents is never invoked.

File: packages/agent/src/dkg-agent-lifecycle.ts (hunk 3 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-lifecycle.ts]: Addressed in 04dd6272c. The page walk is extracted into bounded-curator-roster-traversal.ts with a discriminated complete | continue | cycle result. The recovery owner now persists a continue cursor, clears it for complete, and clears/restarts on cycle. Tests cover churn-safe short tails, one-peer progression, cancellation, and exhausted-cursor restart behavior.

File: packages/agent/src/dkg-agent-cg-resolve.ts (hunk 4 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-cg-resolve.ts]: Addressed in 04dd6272c. Added a positive wallet-curator registry fallback test whose bounded page returns peer-curator while findAgents throws. It verifies registry provenance, a limit: 1 request, and zero rich-discovery calls.

File: packages/agent/src/agent-peer-discovery.ts (hunk 1 above)
Comment (branarakic) [reply packages/agent/src/agent-peer-discovery.ts]: Follow-up in 36acaefe4381b0f962bd84bd9c0f04835499c547: the unknown-to-page decoder now snapshots the bounded array length and narrows/copies each indexed element. It never invokes a provider-supplied iterator. The iterator regression fails on 04dd627 and passes here; null, missing, incorrectly typed, oversized, duplicate, and non-monotonic pages are tested directly as unknown values without casts. The page-contract module has 100% scoped coverage across all four metrics.

File: packages/agent/src/dkg-agent-lifecycle.ts (hunk 3 above)
Comment (branarakic) [reply packages/agent/src/dkg-agent-lifecycle.ts]: Follow-up in 36acaefe4381b0f962bd84bd9c0f04835499c547 preserves the explicit complete/continue/cycle model from 04dd627 and its owner-controlled cursor reset. The existing overflowed/nextPageAfterPeerId SDK fields are retained as compatibility projections because they are already present on testnet-canary. The owner prefers the explicit traversal when available and still handles custom implementations returning the prior shape. The legacy partial-roster regression fails on 04dd627 by incorrectly entering confirmed backoff; exhausted-cursor restart also fails there. Both pass here, alongside a full-cycle test that returns to a newly populated first peer. The empty-cursor wrap performs a fresh bounded first-page probe with cancellation/current-generation checks; a nonempty tail remains cycle and cannot certify completeness. Final validation: 345 tests in 11 suites, agent build/public and strict fixture types, lint, 1,801-file inventory, SPARQL lint, and merge compatibility all pass. Both decoder and traversal modules have 100% scoped statement/branch/function/line coverage.

File: packages/agent/src/dkg-agent-lifecycle.ts [hunk 5]
```diff
@@ -7288,7 +7293,10 @@ export class LifecycleSyncMethods extends DKGAgentBase {
     curatorIsLocal: boolean;
     legacyTripleResolved: boolean;
     lookupFailed?: boolean;
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-lifecycle.ts]: 🟡 Issue: Represent the resolved roster only once

**What's wrong**
The new abstraction duplicates its central data and relies on an undocumented equality invariant. This weakens the boundary instead of making traversal state canonical, and forces downstream code to reconstruct meaning from optional fields and booleans.

**Example**
The return type permits `{ peerIds: ['peer-a'], rosterTraversal: { status: 'continue', peerIds: ['peer-b'], nextAfterPeerId: 'peer-b' }, ... }`. SWM would transport to `peer-a` while advancing according to metadata for `peer-b`. Even if current production construction keeps them equal, the type makes that invariant implicit and fragile for mocks and future branches.

**Suggested direction**
Replace the optional nested traversal plus duplicated top-level data with a single discriminated resolution model. SWM should exhaustively consume that model rather than combining fields from two independently valid shapes.

**For Agents**
Redesign the result type in `dkg-agent-lifecycle.ts` so the peer list has one owner. Prefer a discriminated union for bounded, legacy, local, and failed resolution, or flatten traversal status/cursor alongside the single top-level list. Update the SWM consumer and test helper, preserving complete/continue/cycle behavior. Add a type-level fixture proving contradictory peer lists are unrepresentable and retain behavioral coverage for each traversal state.

File: packages/agent/src/bounded-curator-roster-traversal.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+  readAgentPeerPage,
+  validateAgentPeerPageRequest,
+  type AgentPeerPageRequest,
+} from './agent-peer-discovery.js';
+
+interface BoundedCuratorRosterProvider {
```
Comment (otReviewAgent) [inline packages/agent/src/bounded-curator-roster-traversal.ts]: 🟡 Issue: Use the exported discovery contract at the production boundary

**What's wrong**
The canonical interface is currently bypassed by two near-duplicate, weaker structural types. This adds indirection, permits invalid implementations in normal TypeScript code, and creates multiple definitions that can drift as the protocol evolves.

**Example**
A provider whose method returns `Promise<number>` satisfies `BoundedCuratorRosterProvider` because `number` is assignable to `unknown`, although it does not implement `AgentPeerDiscovery`. The error is deferred to runtime despite the PR introducing a public interface specifically to reject such providers statically.

**Suggested direction**
Make `AgentPeerDiscovery` the sole provider abstraction. Runtime validation does not require weakening or duplicating the compile-time contract.

**For Agents**
Delete `BoundedCuratorRosterProvider` and the inline discovery shape in `readAgentPeerPage`; type both parameters as `AgentPeerDiscovery`. Keep `validateAgentPeerPage` around the awaited value for runtime defense against untyped integrations. Adjust malformed-provider tests to call the validator directly or cross the boundary with an explicit unsafe cast, while the type fixture proves malformed providers cannot satisfy the production helper signature.

File: packages/agent/src/dkg-agent-lifecycle.ts [hunk 7]
```diff
@@ -7353,18 +7348,34 @@ export class LifecycleSyncMethods extends DKGAgentBase {
...
           return { peerIds: [], lookupFailed: true };
         }
       };
       let resolution = await resolve();
       assertCurrent();
-      if (resolution.peerIds.length === 0) {
+      // A missing/broken bounded capability is a lookup failure, never an
+      // invitation to invoke the legacy rich-profile fallback via refresh.
+      if (resolution.peerIds.length === 0
+        && resolution.rosterTraversal?.status !== 'cycle'
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-lifecycle.ts]: 🟡 Issue: Collapse the callback-and-mode plumbing into one resolution boundary

**What's wrong**
A single recovery policy is threaded through the system using two separate extension mechanisms: an optional callback and an optional string mode. This scatters feature-specific branching through general refresh and curator-resolution APIs and makes the policy difficult to discover or evolve.

**Example**
Adding another registry-resolution policy would require coordinating the lifecycle closure, the refresh-options callback contract, the generic resolver's string union and branch, and the public wrapper. The bounded policy is therefore spread across three modules rather than owned by one resolution boundary.

**Suggested direction**
Introduce one explicit bounded refresh/resolution operation, or inject a typed strategy at a single stable boundary. Avoid combining a one-off callback escape hatch with a string mode on the general-purpose resolver.

**For Agents**
Refactor the flow spanning `dkg-agent-lifecycle.ts`, `curator-meta-refresh.ts`, and `dkg-agent-cg-resolve.ts`. Select the bounded resolution strategy once in a dedicated coordinator or explicit resolver function, then invoke a lower-level refresh operation with the resolved peer. Preserve legacy resolution for existing callers and bounded-first-page behavior for recovery. Tests should show both paths use their intended registry lookup without requiring callback or string-mode plumbing through unrelated APIs.

File: packages/agent/src/dkg-agent-swm-host.ts [hunk 8]
```diff
@@ -5880,12 +5884,20 @@ export class SwmHostModeMethods extends DKGAgentBase {
     if (resolutionSucceeded) {
       this.vmReconcileCuratorPeersByCg.delete(localCgId);
       this.vmReconcileCuratorPageCursorByCg.delete(localCgId);
-    } else if (curatorResolution.nextPageAfterPeerId) {
+    } else if (curatorRosterTraversal?.status === 'continue') {
       this.vmReconcileCuratorPageCursorByCg.delete(localCgId);
       this.vmReconcileCuratorPageCursorByCg.set(
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-swm-host.ts:5893]: 🟡 Issue: Test the recovery owner's cursor restart after a completed page cycle

**What's wrong**
The page helper is tested to return `status: 'cycle'`, but no test verifies that the SWM recovery owner consumes that status by clearing its persisted cursor. A regression in this branch could leave recovery repeatedly querying an exhausted tail or incorrectly handling roster evidence while all current tests remain green.

**Example**
Drive `recoverVmReconcileBatch` through three resolutions: `continue`, then tail `cycle`, then a fresh `complete` first page. Assert the resolver receives `afterPeerId` as `[undefined, previousCursor, undefined]` and that the tail is not treated as authoritative absence evidence.

**Suggested direction**
Add an owner-level regression test covering cursor advancement, tail exhaustion, cursor clearing, and the subsequent fresh first-page query.

**For Agents**
In `core-fills-gap.test.ts` or another SWM-host test, exercise `recoverVmReconcileBatch` with mocked traversal results that transition from `continue` to `cycle` to a fresh first page. Prove that the host clears its stored cursor after `cycle`, restarts at the beginning on the next pass, and preserves unconfirmed-roster semantics until a fresh complete page is observed.

File: packages/agent/src/dkg-agent-lifecycle.ts (hunk 5 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-lifecycle.ts]: Addressed in 5809ba4ff. CuratorPeerIdsResolution now has one top-level peerIds roster and a discriminated rosterStatus union for complete, continue, and cycle. The exported type fixture proves a second rosterTraversal peer list is rejected, and SWM consumes the states exhaustively.

File: packages/agent/src/bounded-curator-roster-traversal.ts (hunk 6 above)
Comment (branarakic-agent) [reply packages/agent/src/bounded-curator-roster-traversal.ts]: Addressed in 5809ba4ff. Both page readers now accept the exported AgentPeerDiscovery contract; the duplicate weak provider shapes were removed. Malformed-runtime coverage still exercises the validator directly without weakening the production boundary.

File: packages/agent/src/dkg-agent-lifecycle.ts (hunk 7 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-lifecycle.ts]: Addressed in 5809ba4ff. The callback and registryLookup mode were removed from the public flow. Bounded selection now has an explicit resolveBoundedCuratorSyncPeer boundary, and refresh of that resolved peer uses runCuratorMetaRefreshFromPeer.

File: packages/agent/src/dkg-agent-swm-host.ts (hunk 8 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-swm-host.ts:5893]: Addressed in 5809ba4ff. The recovery-owner test now drives continue, cycle, and fresh complete responses and asserts page cursors [undefined, previousCursor, undefined]. It also verifies the bounded intermediate rosters stay unconfirmed until the complete roster is accepted.

File: packages/agent/src/dkg-agent-lifecycle.ts (hunk 7 above)
Comment (otReviewAgent) [reply packages/agent/src/dkg-agent-lifecycle.ts]: 🟡 Issue: `maxPeerIds` has become a hidden execution-mode switch scattered through lifecycle resolution

**What's wrong**
The option no longer merely limits a result; its presence changes the discovery capability, failure semantics, metadata-refresh resolver, and returned state shape. Encoding that mode through optionality and checking it in multiple distant branches tangles the new feature into a very large general-purpose method.

**Example**
For an empty bounded lookup, line 7372 selects `traverseBoundedCuratorRoster`; line 7413 must independently remember to select `resolveBoundedCuratorSyncPeer` plus `runCuratorMetaRefreshFromPeer`. The unbounded request follows different discovery and refresh semantics through the same options bag.

**Suggested direction**
Make bounded recovery an explicit operation or strategy that owns page lookup, bounded refresh-peer selection, retry, and result construction. This would remove the repeated mode checks and keep bounded policy outside the general resolver.

**For Agents**
Extract an explicit bounded curator-recovery coordinator or dedicated `resolveBoundedCuratorPeerIdsForCg` method, using `BoundedCuratorRosterRequest`. Keep the legacy/unbounded resolver direct. Preserve the bounded path's prohibition on rich-profile fallback, refresh retry, cancellation, and complete/continue/cycle outcomes with the existing bounded discovery tests.

File: packages/agent/src/dkg-agent-swm-host.ts [hunk 9]
```diff
@@ -5803,24 +5804,30 @@ export class SwmHostModeMethods extends DKGAgentBase {
...
-        nextPageAfterPeerId: undefined,
       }));
     if (!isRecoveryCurrent()) return staleRecovery();
     const allResolvedCuratorPeerIds = [...new Set(curatorResolution.peerIds
       .filter((peerId) => peerId && peerId !== this.peerId))]
       .sort((left, right) => left.localeCompare(right));
-    const curatorRosterOverflow = curatorResolution.overflowed === true
+    // Runtime adapter for older custom hosts. The canonical return type below
+    // does not admit pagination without a rosterStatus discriminant.
+    const legacyPagination = curatorResolution as unknown as {
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-swm-host.ts:5819]: 🟡 Issue: The “canonical” roster state still travels through three competing schemas

**What's wrong**
The new discriminated state does not replace the previous model; it is layered beside it. The cast deliberately defeats the exported type, and compatibility branches are now embedded in a 7k-line recovery host. This preserves rather than deletes the representational complexity the new abstraction was meant to solve.

**Example**
A continuing page travels through three shapes: `{ status: 'continue', nextAfterPeerId }`, then `{ rosterStatus: 'continue', overflowed: true, nextPageAfterPeerId }`, then the host reconstructs overflow while also probing a cast legacy cursor.

**Suggested direction**
Use one discriminant and one cursor name end-to-end. Remove the thin `curatorRosterResolution` projection and keep any legacy decoding out of the already-busy recovery orchestration.

**For Agents**
Define one canonical complete/continue/cycle roster-page type in the bounded traversal module and carry it unchanged through lifecycle resolution into SWM recovery. If old custom-host results genuinely require runtime support, normalize `unknown` once at an explicit compatibility boundary. Preserve cursor restart and negative-proof behavior with the existing cycle tests.

File: packages/agent/src/dkg-agent-lifecycle.ts [hunk 10]
```diff
@@ -7353,41 +7394,66 @@ export class LifecycleSyncMethods extends DKGAgentBase {
...
+        try {
+          if (options.maxPeerIds !== undefined) {
+            const refreshPeer = await resolveBoundedCuratorSyncPeer(
+              this,
+              this.preferredSyncPeers,
+              contextGraphId,
+              { signal: options.signal },
+            );
+            assertCurrent();
+            refreshed = await runCuratorMetaRefreshFromPeer(
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-lifecycle.ts:7421]: 🟡 Issue: The successful bounded metadata-refresh path is not exercised

**What's wrong**
This is the new recovery path that connects bounded discovery to metadata installation, but the added tests only cover its no-peer early return. Because curator recovery is user-facing and affects whether missing graph data can be recovered, independently tested components do not verify that this orchestration actually performs the refresh.

**Example**
A regression that passes the wrong peer ID to runCuratorMetaRefreshFromPeer—or makes that wrapper always return false—would leave the current tests green. A regression test could start with an empty bounded roster, supply a bootstrap/projection refresh peer, and assert that metadata fetch runs against that peer before the roster is queried again.

**Suggested direction**
Add an integration-style regression test that drives the nonempty refresh-peer branch through resolveCuratorPeerIdsForCg instead of testing the resolver and refresh components separately.

**For Agents**
In bounded-curator-discovery.test.ts, cover resolveCuratorPeerIdsForCg with maxPeerIds, an initially empty page, and a resolvable bootstrap/projection peer. Exercise the real refresh coordinator enough to prove the selected peer is used, rich discovery remains unused, and the bounded roster is re-read afterward.

File: packages/agent/src/dkg-agent-swm-host.ts (hunk 9 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-swm-host.ts:5819]: Addressed in 5809ba4ff. BoundedCuratorRosterResolution is now the single lifecycle and SWM-recovery representation with one peer list, one rosterStatus discriminant, and one cursor name. Legacy result probing was removed from the recovery host; bounded versus rich registry lookup is selected at an explicit resolver boundary.

File: packages/agent/src/dkg-agent-lifecycle.ts (hunk 10 above)
Comment (branarakic-agent) [reply packages/agent/src/dkg-agent-lifecycle.ts:7421]: Addressed in 5809ba4ff. The bounded-curator integration coverage now drives an initially empty roster through a resolved bounded refresh peer, verifies runCuratorMetaRefreshFromPeer uses that exact peer, confirms rich discovery is never called, and asserts the bounded roster is reread before resolution completes.
